src/atoms.rs

```diff
@@ -673,95 +1232,315 @@ impl<R: Read + Seek> ReadBox<&mut BufReader<R>> for VmhdBox {
             }
+            start += s - HEADER_SIZE;
+            entries.push(DumpBox {name, size: s})
         }
-        skip(reader, current, size);
+        skip_read(reader, current, size);
 
         Ok(StsdBox {
             version,
             flags,
+            entry_count,
+            entries,
         })
     }
 }
 
-fn skip<R: Read + Seek>(reader: &mut BufReader<R>, current: u64, size: u32) {
+impl<W: Write> WriteBox<&mut BufWriter<W>> for StsdBox {
+    fn write_box(&self, _writer: &mut BufWriter<W>) -> Result<u64> {
+        // TODO
+        Ok(0)
+    }
+}
+
+fn skip_read<R: Read + Seek>(reader: &mut BufReader<R>, current: u64, size: u64) {
     let after = reader.seek(SeekFrom::Current(0)).unwrap();
-    let remaining_bytes = (size as u64 - (after - current)) as i64;
+    let remaining_bytes = (size - (after - current)) as i64;
     reader.seek(SeekFrom::Current(remaining_bytes - HEADER_SIZE as i64)).unwrap();
 }
+
+fn skip_write<W: Write>(writer: &mut BufWriter<W>, size: u64) {
+    for _ in 0..size {
+        writer.write_u8(0).unwrap();
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::Cursor;
+
+    #[test]
+    fn test_fourcc() {
+        let ftyp_fcc = 0x66747970;
+        let ftyp_value = FourCC::from(ftyp_fcc);
+        assert_eq!(ftyp_value.value, "ftyp");
+        let ftyp_fcc2 = ftyp_value.into();
+        assert_eq!(ftyp_fcc, ftyp_fcc2);
+    }
+
+    #[test]
+    fn test_ftyp() {
+        let src_box = FtypBox {
+            major_brand: FourCC { value: String::from("isom") },
+            minor_version: 0,
+            compatible_brands: vec![
+                FourCC { value: String::from("isom") },
+                FourCC { value: String::from("iso2") },
+                FourCC { value: String::from("avc1") },
+                FourCC { value: String::from("mp41") },
+            ]
+        };
+        let mut buf = Vec::new();
+        {
+            let mut writer = BufWriter::new(&mut buf);
+            src_box.write_box(&mut writer).unwrap();
+        }
+        assert_eq!(buf.len(), src_box.box_size() as usize);
+
+        {
+            let mut reader = BufReader::new(Cursor::new(&buf));
+            let header = read_box_header(&mut reader, 0).unwrap();
+            assert_eq!(header.name, BoxType::FtypBox);
+            assert_eq!(src_box.box_size(), header.size);
+
+            let dst_box = FtypBox::read_box(&mut reader, header.size).unwrap();
+
+            assert_eq!(src_box, dst_box);
+        }
+    }
+
+    #[test]
+    fn test_mvhd() {
+        let src_box = MvhdBox {
+            version: 0,
+            flags: 0,
+            creation_time: 100,
+            modification_time: 200,
+            timescale: 1000,
+            duration: 634634,
+            rate: 0x00010000,
+        };
+        let mut buf = Vec::new();
+        {
+            let mut writer = BufWriter::new(&mut buf);
+            src_box.write_box(&mut writer).unwrap();
+        }
```

Comment:
Question: Why encapsulate these into a new set of braces? Is this a styling preference? I am new to Rust (first project) and haven't come across this yet. Thanks.

Reply:
It is only to avoid lifetime errors.
If they are not encapsulated into the set of braces, the following errors occur.
```
error[E0502]: cannot borrow `buf` as immutable because it is also borrowed as mutable
...

    let mut writer = BufWriter::new(&mut buf);
                                    -------- mutable borrow occurs here
    src_box.write_box(&mut writer).unwrap();
    assert_eq!(buf.len(), src_box.box_size() as usize);
               ^^^ immutable borrow occurs here
```

Reply:
Ah, gotcha. Thanks!